### ISSUE TITLE
Interleavings of upload operations can cause fluctuations of remote consistent lsn

### ISSUE THREAD

**Issue by LizardWizzard** [neondatabase/neon#3959]
See the discussion here: 
https://neondb.slack.com/archives/C033RQ5SPDH/p1680549635065579

TLDR there are several factors. 

Upload operation granularity allows for UploadMetadata operations from different sources to interleave with each other. E g gc issues remove file operation followed by upload metadata operation. At the same time flush layer task can issue layer upload operation followed by metadata upload.

This doesnt play well with the way how we update metadata file on disk. There is no mechanism that prevents "lost update" anomaly (random link https://dzone.com/articles/what-is-a-lost-update-in-database-systems). 

To partially resolve the issue we might implement some transactional API for RemoteTimelineClient to submit such changes as one operation.

This still leaves the question about on disk metadata file open. We can consider using CAS-like approach for metadata changes.

Please look at the linked slack discussion. There are some ascii diagrams that bring more context.
